v21.1/inverted-indexes.md

```diff
@@ -143,9 +143,23 @@ If you require comparisons using [`<`](functions-and-operators.html#supported-op
 - "is contained by": [`<@`](functions-and-operators.html#supported-operations)
 - "contains": [`@>`](functions-and-operators.html#supported-operations)
 
+## Partial inverted indexes
+
+<span class="version-tag">New in v21.1:</span> You can create a [partial](partial-indexes.html) inverted index, an inverted index on a subset of `JSON`, `ARRAY`, or geospatial container column data. Just like partial indexes that use non-container datatypes, create a partial inverted index by including a clause that evaluates to true on a boolean predicate, like a `WHERE` clause.
+
+{% include copy-clipboard.html %}
+~~~ sql
+CREATE TABLE test (
+  id INT,
+  data JSONB,
+  INVERTED INDEX idx_data(data) WHERE id > 10
+);
+~~~
+
 ## Known limitations
 
-CockroachDB does not support partitioning inverted indexes. For details, see [tracking issue](https://github.com/cockroachdb/cockroach/issues/43643).
+- CockroachDB does not support partitioning inverted indexes. For details, see [tracking issue](https://github.com/cockroachdb/cockroach/issues/43643).
+- CockroachDB does not support [index hinting](table-expressions.html#force-index-selection) for partial inverted indexes.
```

Comment:
This is supported too! Did you run into a case where it didn't work as expected?

Reply:
Hmm, this is on the latest `release-21.1`:

```
CREATE TABLE users (
    profile_id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
    last_updated TIMESTAMP DEFAULT now(),
    user_profile JSONB,
    INVERTED INDEX user_details (user_profile)
  );
INSERT INTO users (user_profile) VALUES
    ('{"first_name": "Lola", "last_name": "Dog", "location": "NYC", "online" : true, "friends" : 547}'),
    ('{"first_name": "Ernie", "status": "Looking for treats", "location" : "Brooklyn"}'),
    ('{"first_name": "Carl", "last_name": "Kimball", "location": "NYC", "breed": "Boston Terrier"}'
  );
CREATE INVERTED INDEX idx_online_users ON users(user_profile) WHERE user_profile -> 'online' = 'true';
```

Then try to query the partial inverted index:

```
SELECT * FROM users@idx_online_users;
```

which gives:

```
ERROR: index "idx_online_users" is a partial inverted index and cannot be used for this query
```

And:

```
SELECT * FROM users@{FORCE_INDEX=idx_online_users};
```

which also gives:

```
ERROR: index "idx_online_users" is a partial inverted index and cannot be used for this query
```

Am I doing something wrong?

Reply:
Your `SELECT` queries don't contain a `WHERE` clause that implies the partial index predicate. You're querying for all the rows in the table, but `idx_online_users` doesn't have an entry for every row in the table, so it can't answer your query.

This behavior is the same for non-inverted partial indexes too:

```sql
CREATE TABLE t (a INT, INDEX idx (a) WHERE a > 0);
CREATE TABLE

SELECT * FROM t@idx;
ERROR: index "idx" is a partial index that does not contain all the rows needed to execute this query

SELECT * FROM t@idx WHERE a > 0;
  a
-----
(0 rows)
```

Also, we don't support forcing an unconstrained scan on an inverted index, partial or not:

```sql
CREATE TABLE t (j JSON, INVERTED INDEX idx (j));
CREATE TABLE

SELECT * FROM t@idx;
ERROR: index "idx" is inverted and cannot be used for this query

SELECT * FROM t@idx WHERE j @> '1';
  j
-----
(0 rows)
```

The scan would be unconstrained because there's no `WHERE` clause that can be used to constrain the index scan. In theory, we could support this, but in practice it doesn't make much sense - an inverted index can have many more entries than rows, so if you're querying all rows then scanning a non-inverted index, like the primary index, is always better than scanning an inverted index.

So in order to force a **partial inverted index** you'd need to include a `WHERE` filter that (1) implies the partial index predicate and (2) constrains the inverted index scan. If both are not satisfied, you'll get an error.

A final demonstration:

```sql
CREATE TABLE t (j JSON, INVERTED INDEX idx (j) WHERE j->'a' = '1');
CREATE TABLE

-- This WHERE filter satisfies (1), but not (2).
SELECT * FROM t@idx WHERE j->'a' = '1';
ERROR: index "idx" is a partial inverted index and cannot be used for this query

-- This WHERE filter satisfies (2), but not (1).
SELECT * FROM t@idx WHERE j->'b' = '2';
ERROR: index "idx" is a partial inverted index and cannot be used for this query

-- This WHERE filter satisfies (1) and (2).
SELECT * FROM t@idx WHERE j->'a' = '1' AND j->'b' = '2';
  j
-----
(0 rows)
```

**TLDR;** If you update your query to something like below, forcing the index should work:

```sql
SELECT * FROM users@idx_online_users WHERE user_profile->'online' = 'true' AND user_profile->'location' = '"NYC"'
```

Reply:
Ah, ok. I think I get it now.

Reply:
Removed the bogus limitation, and added examples for both partial and partial inverted indexes to the index forcing topic.